### PR TITLE
disable listeners if proxy is configured

### PR DESCRIPTION
**Pull request by sickyoon**
disable listeners if proxy is configured
https://github.com/anacrolix/torrent/issues/359

### CODE REVIEW

**Comment by sickyoon** [comment]
dependency changes (go.mod) are there for upstream project.
I will remove them once you review this PR.

File: socket.go [hunk 1]
```diff
@@ -55,11 +56,14 @@ func listenTcp(network, address, proxyURL string) (s socket, err error) {
 	// If we don't need the proxy - then we should return default net.Dialer,
 	// otherwise, let's try to parse the proxyURL and return proxy.Dialer
 	if len(proxyURL) != 0 {
+
+		dl := disabledListener{l}
+
 		// TODO: The error should be propagated, as proxy may be in use for
 		// security or privacy reasons. Also just pass proxy.Dialer in from
 		// the Config.
 		if dialer, err := getProxyDialer(proxyURL); err == nil {
```
Comment (anacrolix) [inline socket.go]: Wouldn't it be more appropriate instead to not ignore errors here? I'm not sure about how listening should be handled when proxies are in use.

Comment (sickyoon) [reply socket.go]: Yeah we should not ignore url parse error here. My understanding is that listening can't be handled through socks5 proxy and should be disabled when proxy is in effect.

Comment (anacrolix) [reply socket.go]: That makes sense. Do you want to make a PR for handling the error separately?

Comment (anacrolix) [reply socket.go]: I think @elgatito would know more about this, did you want to make any recommendations, Denis?

Comment (elgatito) [reply socket.go]: @anacrolix When I initially added proxy support, I've tested cases when specifying non-working proxy or proxy with wrong credentials. In both cases downloading was not working (as it should be).
Testing can by done with installing dante proxy (pretty simple to install) with authorization, and proxying through it.

Comment (sickyoon) [reply socket.go]: @elgatito perhaps listeners were added after your implementation but torrent can definitely downloaded with invalid proxy. You can try to test it yourself. @anacrolix I added error handling part to my PR? could you review and merge?

Comment (anacrolix) [reply socket.go]: Erroring on a bad proxy seems like the more correct behaviour, for privacy and security reasons.

**Pull request merged**